aspnetcore/fundamentals/host/hosted-services/samples/2.x/BackgroundTasksSample-GenericHost/Program.cs

```diff
@@ -0,0 +1,109 @@
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using BackgroundTasksSample.Services;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Logging;
+
+namespace BackgroundTasksSample
+{
+    public class Program
+    {
+        public static async Task Main(string[] args)
+        {
+            var host = new HostBuilder()
+                .ConfigureLogging((hostContext, config) =>
+                {
+                    config.AddConsole();
+                    config.AddDebug();
+                })
+                .ConfigureAppConfiguration((hostContext, config) =>
+                {
+                    config.AddEnvironmentVariables();
+                    config.AddJsonFile("appsettings.json", optional: true);
+                    config.AddJsonFile($"appsettings.{hostContext.HostingEnvironment.EnvironmentName}.json", optional: true);
+                    config.AddCommandLine(args);
+                })
+                .ConfigureServices((hostContext, services) =>
+                {
+                    services.AddLogging();
+
+                    #region snippet1
+                    services.AddHostedService<TimedHostedService>();
+                    #endregion
+
+                    #region snippet2
+                    services.AddHostedService<ConsumeScopedServiceHostedService>();
+                    services.AddScoped<IScopedProcessingService, ScopedProcessingService>();
+                    #endregion
+
+                    #region snippet3
+                    services.AddHostedService<QueuedHostedService>();
+                    services.AddSingleton<IBackgroundTaskQueue, BackgroundTaskQueue>();
+                    #endregion
+                })
+                .UseConsoleLifetime()
+                .Build();
+
+            using (host)
+            {
+                // Start the host
+                await host.StartAsync();
+
+                // Monitor for new background queue work items
+                StartMonitorLoop(host);
+
+                // Wait for the host to shutdown
+                await host.WaitForShutdownAsync();
+            }
+        }
+        
+        private static void StartMonitorLoop(IHost host)
+        {
+            var loggerFactory = host.Services.GetRequiredService<ILoggerFactory>();
+            var backgroundTaskQueue = host.Services.GetRequiredService<IBackgroundTaskQueue>();
+            var backgroundTaskLogger = loggerFactory.CreateLogger<IBackgroundTaskQueue>();
+            var applicationLifetime = host.Services.GetRequiredService<IApplicationLifetime>();
+
+            // Run a console user input loop in a background thread
+            Task.Run(() => MonitorLoop(backgroundTaskQueue, backgroundTaskLogger, applicationLifetime.ApplicationStopping));
```

Comment:
@Tratcher Silly question (i.e., my standard type of question :grinning:): The new type wouldn't get consumed from DI anyplace else. Feels like it would just add unnecessary complexity. Is there some other reason to add the type?

Reply:
DI would do the construction for you. Lines 65-68 are duplicating functionality from DI.

Reply:
Ok ... I see ... I'll give it a go.

Reply:
`host.Services.GetRequiredService<MonitorLoop>()`

Reply:
Thx!